common/tasks/multi_destination_scheduler.go

```diff
@@ -0,0 +1,241 @@
+	stopFn       context.CancelFunc
+	limitersLock sync.RWMutex
+	limiters     map[K]*limiter[T]
+}
+
+func NewMultiDestinationScheduler[K comparable, T Task](options MultiDestinationSchedulerOptions[K, T]) *MultiDestinationScheduler[K, T] {
+	stopCtx, stopFn := context.WithCancel(context.Background())
+	return &MultiDestinationScheduler[K, T]{
+		options:  options,
+		stopCtx:  stopCtx,
+		stopFn:   stopFn,
+		limiters: make(map[K]*limiter[T]),
+	}
+}
+
+func (*MultiDestinationScheduler[K, T]) Start() {
+	// noop
+}
+
+// Stop signals running tasks to stop, aborts any pending tasks and waits up to a minute for all running tasks to
+// complete.
+func (s *MultiDestinationScheduler[K, T]) Stop() {
+	s.stopFn()
+	s.abortBufferedTasks()
+	if success := common.AwaitWaitGroup(&s.waitGroup, time.Minute); !success {
+		s.options.Logger.Warn("MultiDestinationScheduler timed out waiting for tasks to complete")
+	} else {
+		s.options.Logger.Debug("MultiDestinationScheduler drained")
+	}
+}
+
+func (s *MultiDestinationScheduler[K, T]) abortBufferedTasks() {
+	s.limitersLock.Lock()
+	defer s.limitersLock.Unlock()
+
+	for _, lim := range s.limiters {
+		lim.Lock()
+		for _, task := range lim.buffer {
+			task.Abort()
+		}
+		lim.buffer = nil
+		lim.Unlock()
+	}
+}
+
+// Submit is not implemented as it is never invoked.
+func (*MultiDestinationScheduler[K, T]) Submit(T) {
+	panic("unimplemented")
+}
+
+// TrySubmit submits a task for processing unless the scheduler is shut down.
+func (s *MultiDestinationScheduler[K, T]) TrySubmit(task T) bool {
+	// Add to the wait group before checking if we're stopped to prevent Stop() from returning early if TrySubmit is
+	// called concurrently.
+	s.waitGroup.Add(1)
+	if s.stopCtx.Err() != nil {
+		s.waitGroup.Done()
+		return false
+	}
+
+	if lim := s.limiterForTask(task); lim != nil {
+		go func() {
+			defer s.waitGroup.Done()
+			s.executeUntilReleased(task, lim)
+		}()
+	} else {
+		s.waitGroup.Done()
+	}
+	return true
+}
+
+// limiterForTask gets or creates a limiter for the task's destination.
+// If the limiter has hit the concurrency limit, returns nil and buffers the task to be processed later.
+func (s *MultiDestinationScheduler[K, T]) limiterForTask(task T) *limiter[T] {
+	key := s.options.KeyFn(task)
+	s.limitersLock.RLock()
+	lim, ok := s.limiters[key]
+	s.limitersLock.RUnlock()
+	if !ok {
+		s.limitersLock.Lock()
+		// Check again in case the limiter map was populated between releasing and aquiring the lock.
+		if lim, ok = s.limiters[key]; !ok {
+			lim = newLimiter[T](s.options.LimiterOptionsFn(key))
+			s.limiters[key] = lim
+		}
+		s.limitersLock.Unlock()
+	}
+
+	lim.Lock()
+	defer lim.Unlock()
+	if !lim.reserve() {
+		lim.enqueue(task)
+		return nil
+	}
+	return lim
+}
+
+// executeUntilReleased execute tasks starting from the given task using the provided limiter.
+// Continues as long as it has tasks to dequeue.
+func (s *MultiDestinationScheduler[K, T]) executeUntilReleased(task T, lim *limiter[T]) {
```

Comment:
optional nit: I think `func (l *limiter[T]) executeUntilBufferIsEmpty(ctx context.Context, task T)` is a more natural signature here for two reasons:
1) The scheduler isn't really involved here other than providing the context, so I don't think it should be the method receiver, especially because, at this point, there's nothing "multi-destination" specific about this logic.
2) The term "released", to me, doesn't convey that this continues executing tasks until there are no more tasks in the buffer. It sounds like some other party is telling the executor it can now stop, which is only true in the case of a canceled context, not when the task buffer is empty, and the former is implied if the context is part of the signature.